staging/src/k8s.io/apiserver/pkg/util/peerproxy/peer_discovery.go

```diff
@@ -89,63 +96,85 @@ func (h *peerProxyHandler) syncPeerDiscoveryCache(ctx context.Context) error {
 		if err != nil {
 			fetchDiscoveryErr = err
 		}
-
-		if discoveryInfo != nil {
-			newCache[l.Name] = discoveryInfo
+		// Only add if there is at least one GVR or group
+		if len(discoveryEntry.GVRs) > 0 || len(discoveryEntry.GroupDiscovery) > 0 {
+			newCache[l.Name] = discoveryEntry
 		}
 	}
 
 	// Overwrite cache with new contents.
 	h.peerDiscoveryInfoCache.Store(newCache)
+
+	if len(newCache) != 0 {
+		// After updating peer discovery cache, invalidate discovery manager cache.
+		for _, cb := range h.cacheInvalidationCallbacks {
+			cb()
+		}
+	}
 	return fetchDiscoveryErr
 }
 
-func (h *peerProxyHandler) fetchNewDiscoveryFor(ctx context.Context, serverID string, holderIdentity string) (map[schema.GroupVersionResource]bool, error) {
+func (h *peerProxyHandler) fetchNewDiscoveryFor(ctx context.Context, serverID string) (PeerDiscoveryCacheEntry, error) {
 	hostport, err := h.hostportInfo(serverID)
 	if err != nil {
-		return nil, fmt.Errorf("failed to get host port info from identity lease for server %s: %w", serverID, err)
+		return PeerDiscoveryCacheEntry{}, fmt.Errorf("failed to get host port info from identity lease for server %s: %w", serverID, err)
 	}
 
 	klog.V(4).Infof("Proxying an agg-discovery call from %s to %s", h.serverID, serverID)
-	servedResources := make(map[schema.GroupVersionResource]bool)
+	gvrMap := make(map[schema.GroupVersionResource]bool)
 	var discoveryErr error
 	var discoveryResponse *apidiscoveryv2.APIGroupDiscoveryList
 	discoveryPaths := []string{"/api", "/apis"}
+	groupMap := make(map[string]apidiscoveryv2.APIGroupDiscovery)
 	for _, path := range discoveryPaths {
 		discoveryResponse, discoveryErr = h.aggregateDiscovery(ctx, path, hostport)
-		if err != nil {
-			klog.ErrorS(err, "error querying discovery endpoint for serverID", "path", path, "serverID", serverID)
+		if discoveryErr != nil {
+			klog.ErrorS(discoveryErr, "error querying discovery endpoint for serverID", "path", path, "serverID", serverID)
 			continue
 		}
 
 		for _, groupDiscovery := range discoveryResponse.Items {
-			groupName := groupDiscovery.Name
-			if groupName == "" {
-				groupName = "core"
-			}
-
 			for _, version := range groupDiscovery.Versions {
 				for _, resource := range version.Resources {
-					gvr := schema.GroupVersionResource{Group: groupName, Version: version.Version, Resource: resource.Resource}
-					servedResources[gvr] = true
+					gvr := schema.GroupVersionResource{
+						Group:    groupDiscovery.Name,
+						Version:  version.Version,
+						Resource: resource.Resource,
+					}
+					gvrMap[gvr] = true
 				}
 			}
+			// Skip core/v1 group from peer-aggregated discovery since its not served from /apis.
+			// We still want to re-route core/v1 requests to the peer, but we don't want it
+			// to appear in the peer-aggregated discovery document.
+			if groupDiscovery.Name == "" {
+				continue
+			}
+			groupMap[groupDiscovery.Name] = groupDiscovery
```

Comment:
doesn't using a map here lose ordering we care about?

Reply:
switched to appending to list directly instead of map.